### PR TITLE
OTLP Trace Listeners report internal metrics

### PR DESCRIPTION
**Pull request by oppegard**
Always report `spans.<handle>.received.total`.

If distributed tracing is disabled at the cluster/tenant level, report `spans.<handle>.discarded`.

If span logs are disabled at the cluster/tentant level, report `spanLogs.<handle>.discarded`.

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java [hunk 1]
```diff
@@ -71,13 +71,50 @@ public static boolean isFeatureDisabled(Supplier<Boolean> featureDisabledFlag, S
...
+   * @param featureDisabledFlag Supplier for feature disabled flag.
+   * @param message             Warning message to log if feature is disabled.
+   * @param discardedCounter    Counter for discarded items.
+   * @param increment           The amount by which the counter will be increased.
+   * @return true if feature is disabled
+   */
+  public static boolean isFeatureDisabled(Supplier<Boolean> featureDisabledFlag,
+                                          String message,
+                                          Counter discardedCounter,
+                                          long increment) {
```
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java:90]: Why we need this ?

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java [hunk 2]
```diff
@@ -100,20 +103,22 @@ public static void exportToWavefront(ExportTraceServiceRequest request,
     }
 
     for (Pair<Span, SpanLogs> spanAndLogs : fromOtlpRequest(request, preprocessor, defaultSource)) {
-      if (wasFilteredByPreprocessor(spanAndLogs._1, spanHandler, preprocessor)) {
-        continue;
-      }
+      Span span = spanAndLogs._1;
+      SpanLogs spanLogs = spanAndLogs._2;
```
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java]: Can we use another "structure", Pair with its `_1` and `_2` makes the code ugly and hard to read.

Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java]: I agree - I was looking for something lightweight to group "related" things together. 

We could switch to Apache Commons' [Pair](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/tuple/Pair.html), which would look like:
```java
Span span = spanAndLogs.left();
SpanLogs spanLogs = spanAndLogs.right();
```

Or, use JDK's `AbstractMap.SimpleEntry<Span, SpanLogs>`:
```java
Span span = spanAndLogs.getKey();
SpanLogs spanLogs = spanAndLogs.getValue();
```

Are either of those preferable? If not, any other ideas? I use this same pattern elsewhere in `OtlpProtobufUtils.java`, such as [grouping counters with the associated object](https://github.com/oppegard/wavefront-proxy/blob/0b10d6904c77064792894cd2ec2f38991fa93a52/proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java#L94-L95) that determines whether or not to increment the counter.

File: proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java (hunk 1 above)
Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java:90]: This was added so I could increment `discardedCounter` by the total number of discarded spans in the two Listeners:
- [OtlpGrpcTraceHandler](https://github.com/oppegard/wavefront-proxy/blob/0b10d6904c77064792894cd2ec2f38991fa93a52/proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpGrpcTraceHandler.java#L126)
- [OtlpHttpHandler](https://github.com/oppegard/wavefront-proxy/blob/0b10d6904c77064792894cd2ec2f38991fa93a52/proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpHttpHandler.java#L122)

Comment (laullon) [reply proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java:90]: No need, this is to reduce the log messages, and this change will increment it.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java (hunk 2 above)
Comment (laullon) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java]: May we can create a simple class with two members.

File: proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java (hunk 1 above)
Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java:90]: I'm not sure how you're suggesting to accurately increment the discarded spans counter. Should I use the existing `isFeatureDisabled()` method that increments the counter by one, and then manually increment the discarded spans counter by `spansCount-1` back in the in listener?

With the new method, I can check if the span feature is disabled and increment the total number of spans that will be discarded in the batch in one go: https://github.com/oppegard/wavefront-proxy/blob/0b10d6904c77064792894cd2ec2f38991fa93a52/proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpGrpcTraceHandler.java#L123-L126.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java (hunk 2 above)
Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java]: Ok, I can create a simple container class to hold the two members.

Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufUtils.java]: Resolved in https://github.com/oppegard/wavefront-proxy/pull/11/commits/f09c842b56a7ee416434608bcd7ab9adec4f6cf4

File: proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java (hunk 1 above)
Comment (laullon) [reply proxy/src/main/java/com/wavefront/agent/listeners/FeatureCheckUtils.java:90]: ok, use your method.

I'll review/recode the `isFeatureDisabled` method on the future.

**Review by laullon — APPROVED** [review]

**Pull request merged**